### ISSUE TITLE
Does not recover from error watching kubernetes events on startup

### ISSUE THREAD

**Issue by simt2** [containous/traefik#575]
I'm running a deployment of traefik on kubernetes/GKE with two replicas:

``` yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: traefik
  labels:
    app: traefik
spec:
  replicas: 2
  template:
    metadata:
      name: traefik
      labels:
        app: traefik
    spec:
      terminationGracePeriodSeconds: 60
      containers:
      - image: traefik
        name: traefik
        imagePullPolicy: Always
        ports:
        - containerPort: 80
        - containerPort: 443
        - containerPort: 8080
        args:
        - --web
        - --kubernetes
        - --logLevel=DEBUG
```

Sometimes, when creating an new cluster on GKE and immediately creating above deployment and services plus their ingresses, one of the two replicas ends up without paths and backends while the other boots fine. (I guess this happens because the master is not fully available yet, but that's another issue.) Log of the broken traefik instance:

```
time="2016-08-01T05:34:06Z" level=info msg="Traefik version v1.0.1 built on 2016-07-19_03:07:34PM"
time="2016-08-01T05:34:06Z" level=debug msg="Global configuration loaded {\"GraceTimeOut\":10,\"Debug\":false,\"AccessLogsFile\":\"\",\"TraefikLogsFile\":\"\",\"LogLevel\":\"DEBUG\",\"EntryPoints\":{\"http\":{\"Network\":\"\",\"Address\":\":80\",\"TLS\":null,\"Redirect\":null}},\"Constraints\":[],\"ACME\":null,\"DefaultEntryPoints\":[\"http\"],\"ProvidersThrottleDuration\":2000000000,\"MaxIdleConnsPerHost\":200,\"Retry\":null,\"Docker\":null,\"File\":null,\"Web\":{\"Address\":\":8080\",\"CertFile\":\"\",\"KeyFile\":\"\",\"ReadOnly\":false},\"Marathon\":null,\"Consul\":null,\"ConsulCatalog\":null,\"Etcd\":null,\"Zookeeper\":null,\"Boltdb\":null,\"Kubernetes\":{\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Endpoint\":\"http://127.0.0.1:8080\",\"DisablePassHostHeaders\":false,\"Namespaces\":null}}"
time="2016-08-01T05:34:06Z" level=info msg="Preparing server http &{Network: Address::80 TLS:<nil> Redirect:<nil>}"
time="2016-08-01T05:34:06Z" level=info msg="Starting provider *main.WebProvider {\"Address\":\":8080\",\"CertFile\":\"\",\"KeyFile\":\"\",\"ReadOnly\":false}"
time="2016-08-01T05:34:06Z" level=info msg="Starting provider *provider.Kubernetes {\"Watch\":true,\"Filename\":\"\",\"Constraints\":[],\"Endpoint\":\"http://127.0.0.1:8080\",\"DisablePassHostHeaders\":false,\"Namespaces\":null}"
time="2016-08-01T05:34:06Z" level=info msg="Starting server on :80"
time="2016-08-01T05:34:06Z" level=debug msg="Kubernetes token: secret-token"
time="2016-08-01T05:34:06Z" level=debug msg="Kubernetes CA cert: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
time="2016-08-01T05:34:06Z" level=debug msg="Kubernetes endpoint: https://10.59.240.1:443"
time="2016-08-01T05:36:14Z" level=error msg="Error watching kubernetes events: failed to create watch: failed to do version request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : failed to create request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : [Get https://10.59.240.1:443/apis/extensions/v1beta1/ingresses: dial tcp 10.59.240.1:443: getsockopt: connection timed out]"
time="2016-08-01T05:36:14Z" level=error msg="Error retrieving ingresses: failed to create ingresses request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : failed to create request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : [Get https://10.59.240.1:443/apis/extensions/v1beta1/ingresses: dial tcp 10.59.240.1:443: getsockopt: connection timed out]"
time="2016-08-01T05:36:14Z" level=error msg="Error starting provider failed to create ingresses request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : failed to create request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : [Get https://10.59.240.1:443/apis/extensions/v1beta1/ingresses: dial tcp 10.59.240.1:443: getsockopt: connection timed out]"
time="2016-08-01T05:36:15Z" level=error msg="Kubernetes connection error failed to create watch: failed to do version request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : failed to create request: GET \"https://10.59.240.1:443/apis/extensions/v1beta1/ingresses\" : [Get https://10.59.240.1:443/apis/extensions/v1beta1/ingresses: dial tcp 10.59.240.1:443: getsockopt: connection timed out], retrying in 686.36285ms"
```

Traefik does not seem to retry this request on its own, nor terminates on the error. Only when I delete and add an ingress in Kubernetes, paths and backends show up in the instances web interface.

```
time="2016-08-01T06:05:15Z" level=debug msg="Received event from kubernetes map[type:DELETED object:map[kind:Ingress apiVersion:extensions/v1beta1 metadata:map[name:app-2uuid selfLink:/apis/extensions/v1beta1/namespaces/default/ingresses/app-2uuid resourceVersion:817 generation:1 labels:map[app:app customer:customer2uuid tier:paid] namespace:default uid:7f787211-57a9-11e6-964c-42010af0001a creationTimestamp:2016-08-01T05:33:38Z annotations:map[kubectl.kubernetes.io/last-applied-configuration:{\"kind\":\"Ingress\",\"apiVersion\":\"extensions/v1beta1\",\"metadata\":{\"name\":\"app-2uuid\",\"creationTimestamp\":null,\"labels\":{\"app\":\"app\",\"customer\":\"customer2uuid\",\"tier\":\"paid\"}},\"spec\":{\"rules\":[{\"host\":\"app2.example.com\",\"http\":{\"paths\":[{\"path\":\"/\",\"backend\":{\"serviceName\":\"app-2uuid\",\"servicePort\":80}}]}}]},\"status\":{\"loadBalancer\":{}}}]] spec:map[rules:[map[http:map[paths:[map[path:/ backend:map[serviceName:app-2uuid servicePort:80]]]] host:app2.example.com]]] status:map[loadBalancer:map[]]]]"
time="2016-08-01T06:05:15Z" level=warning msg="Unknown RuleType ``, falling back to `PathPrefix"
time="2016-08-01T06:05:15Z" level=debug msg="Error loading load balancer method '<nil>' for backend app1.example.com/: Invalid method, using default. Using default wrr."
time="2016-08-01T06:05:15Z" level=debug msg="Configuration received from provider kubernetes: {\"backends\":{\"app1.example.com/\":{\"servers\":{\"http://10.56.1.5:8080\":{\"url\":\"http://10.56.1.5:8080\",\"weight\":1}},\"loadBalancer\":{\"method\":\"wrr\"}}},\"frontends\":{\"app1.example.com/\":{\"entryPoints\":[\"http\"],\"backend\":\"app1.example.com/\",\"routes\":{\"/\":{\"rule\":\"PathPrefix:/\"},\"app1.example.com\":{\"rule\":\"Host:app1.example.com\"}},\"passHostHeader\":true,\"priority\":0}}}"
time="2016-08-01T06:05:15Z" level=debug msg="Last kubernetes config received more than 2s, OK"
time="2016-08-01T06:05:15Z" level=debug msg="Creating frontend app1.example.com/"
time="2016-08-01T06:05:15Z" level=debug msg="Wiring frontend app1.example.com/ to entryPoint http"
time="2016-08-01T06:05:15Z" level=debug msg="Creating route / PathPrefix:/"
time="2016-08-01T06:05:15Z" level=debug msg="Creating route app1.example.com Host:app1.example.com"
time="2016-08-01T06:05:15Z" level=debug msg="Creating backend app1.example.com/"
time="2016-08-01T06:05:15Z" level=debug msg="Creating load-balancer wrr"
time="2016-08-01T06:05:15Z" level=debug msg="Creating server http://10.56.1.5:8080 at http://10.56.1.5:8080 with weight 1"
time="2016-08-01T06:05:15Z" level=info msg="Server configuration reloaded on :80"
time="2016-08-01T06:05:20Z" level=debug msg="Received event from kubernetes map[type:ADDED object:map[spec:map[rules:[map[host:app2.example.com http:map[paths:[map[path:/ backend:map[serviceName:app-2uuid servicePort:80]]]]]]] status:map[loadBalancer:map[]] kind:Ingress apiVersion:extensions/v1beta1 metadata:map[uid:edb2c4fe-57ad-11e6-964c-42010af0001a resourceVersion:821 generation:1 creationTimestamp:2016-08-01T06:05:21Z labels:map[customer:customer2uuid tier:paid app:app] name:app-2uuid namespace:default selfLink:/apis/extensions/v1beta1/namespaces/default/ingresses/app-2uuid]]]"
time="2016-08-01T06:05:20Z" level=warning msg="Unknown RuleType ``, falling back to `PathPrefix"
time="2016-08-01T06:05:20Z" level=warning msg="Unknown RuleType ``, falling back to `PathPrefix"
time="2016-08-01T06:05:20Z" level=debug msg="Error loading load balancer method '<nil>' for backend app1.example.com/: Invalid method, using default. Using default wrr."
time="2016-08-01T06:05:20Z" level=debug msg="Error loading load balancer method '<nil>' for backend app2.example.com/: Invalid method, using default. Using default wrr."
time="2016-08-01T06:05:20Z" level=debug msg="Configuration received from provider kubernetes: {\"backends\":{\"app1.example.com/\":{\"servers\":{\"http://10.56.1.5:8080\":{\"url\":\"http://10.56.1.5:8080\",\"weight\":1}},\"loadBalancer\":{\"method\":\"wrr\"}},\"app2.example.com/\":{\"servers\":{\"http://10.56.0.5:8080\":{\"url\":\"http://10.56.0.5:8080\",\"weight\":1}},\"loadBalancer\":{\"method\":\"wrr\"}}},\"frontends\":{\"app1.example.com/\":{\"entryPoints\":[\"http\"],\"backend\":\"app1.example.com/\",\"routes\":{\"/\":{\"rule\":\"PathPrefix:/\"},\"app1.example.com\":{\"rule\":\"Host:app1.example.com\"}},\"passHostHeader\":true,\"priority\":0},\"app2.example.com/\":{\"entryPoints\":[\"http\"],\"backend\":\"app2.example.com/\",\"routes\":{\"/\":{\"rule\":\"PathPrefix:/\"},\"app2.example.com\":{\"rule\":\"Host:app2.example.com\"}},\"passHostHeader\":true,\"priority\":0}}}"
time="2016-08-01T06:05:20Z" level=debug msg="Last kubernetes config received more than 2s, OK"
time="2016-08-01T06:05:20Z" level=debug msg="Creating frontend app1.example.com/"
time="2016-08-01T06:05:20Z" level=debug msg="Wiring frontend app1.example.com/ to entryPoint http"
time="2016-08-01T06:05:20Z" level=debug msg="Creating route / PathPrefix:/"
time="2016-08-01T06:05:20Z" level=debug msg="Creating route app1.example.com Host:app1.example.com"
time="2016-08-01T06:05:20Z" level=debug msg="Creating backend app1.example.com/"
time="2016-08-01T06:05:20Z" level=debug msg="Creating load-balancer wrr"
time="2016-08-01T06:05:20Z" level=debug msg="Creating server http://10.56.1.5:8080 at http://10.56.1.5:8080 with weight 1"
time="2016-08-01T06:05:20Z" level=debug msg="Creating frontend app2.example.com/"
time="2016-08-01T06:05:20Z" level=debug msg="Wiring frontend app2.example.com/ to entryPoint http"
time="2016-08-01T06:05:20Z" level=debug msg="Creating route / PathPrefix:/"
time="2016-08-01T06:05:20Z" level=debug msg="Creating route app2.example.com Host:app2.example.com"
time="2016-08-01T06:05:20Z" level=debug msg="Creating backend app2.example.com/"
time="2016-08-01T06:05:20Z" level=debug msg="Creating load-balancer wrr"
time="2016-08-01T06:05:20Z" level=debug msg="Creating server http://10.56.0.5:8080 at http://10.56.0.5:8080 with weight 1"
time="2016-08-01T06:05:20Z" level=info msg="Server configuration reloaded on :80"
```

**Comment by emilevauge** [comment]
@errm Do you have an idea of what's happening here?

**Comment by abh** [comment]
Is this fixed by using the standard k8 go client on master?

**Comment by timoreimann** [comment]
This needs to be re-triaged once 1.2.1 lands (which will contain [this important fix](https://github.com/containous/traefik/pull/1331)).

@simt2 do you still have a chance to give it a whirl again once 1.2.1 is out?

**Comment by simt2** [comment]
@timoreimann I certainly will try, but the error is hard to reproduce.

**Comment by timoreimann** [comment]
@simt2 one way to test this is to run Traefik locally and have it connect to your cluster through a local endpoint set up by `kubectl proxy`. (Traefik 1.2.0 added support for external cluster usage.) If you only start the proxy after Traefik has first tried to reach out to the Kubernetes master, it should be pretty easy to see if it works.

If you have an the opportunity to test things with 1.2.1 some time next week, that'd be great. Otherwise, I can also test it myself. Just let me know.

**Comment by simt2** [comment]
@timoreimann I'm happy to help with testing next week, still using traefik in two of my projects! Just waiting for 1.2.1 to land.

**Comment by Regner** [comment]
1.2.1 is out now.

**Comment by timoreimann** [comment]
@simt2 are you able to test again with 1.2.1+?

**Comment by timoreimann** [comment]
This is working fine now according to my local tests.

Closing.
